### PR TITLE
Add GitHub workflow success ratio metrics for inferred build status

### LINKED ISSUE #161: Biomedical research meaningful measures: Table 4 - Version control - Current build status
Measure whether the current build is passing without errors somehow. Perhaps a quick check to GitHub Actions, if a GitHub repo is detected.

### PR DESCRIPTION
**Pull request by d33bs**
# Description

This PR adds metrics which infer the current build status by using the latest workflow runs for the default head branch of a repository hosted on GitHub. I couldn't find an abstraction for this effort so we won't have these details yet for GitLab and similar, though I feel we should reach these eventually via a similar mechanism.

The thought process behind these metrics is: we can't possibly know how to build or safely perform a build for a repository without human understanding efforts, therefore we use a ratio of the repo-defined workflows to better understand what might be happening with builds (as software is commonly built to perform tests or other workflow-related work). This also I feel is related to testing coverage but we should give that particular aspect a deeper dive with other PR's.

Closes #161 
References #162

## What is the nature of your change?

- [ ] Content additions or updates (adds or updates content)
- [ ] Bug fix (fixes an issue).
- [x] Enhancement (adds functionality).
- [ ] Breaking change (these changes would cause existing functionality to not work as expected).

# Checklist

Please ensure that all boxes are checked before indicating that this pull request is ready for review.

- [x] I have read the [CONTRIBUTING.md](CONTRIBUTING.md) guidelines.
- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own contributions.
- [x] I have commented my content, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to related documentation (outside of book content).
- [x] My changes generate no new warnings.
- [x] New and existing tests pass locally with my changes.
- [x] I have added tests that prove my additions are effective or that my feature works.
- [x] I have deleted all non-relevant text in this pull request template.

### CODE REVIEW

File: tests/metrics/test_data.py [hunk 1]
```diff
@@ -692,3 +693,22 @@ def test_fetch_api_data(current_repo):
     assert (
         repo_data["html_url"] == remote_url
     ), "The repo_data URL should match the repository's remote URL."
+
+
+def test_get_github_build_success_ratio():
```
Comment (gwaybio) [inline tests/metrics/test_data.py:698]: the name of this test implies you're only testing success ratio? Oh, I see, this is the name of the function! Although it does look like the function is doing more than getting the success ratio?

Comment (gwaybio) [reply tests/metrics/test_data.py:698]: I'm also wondering if it's worth checking the requests failures? It might not be

**Review by gwaybio — APPROVED** [review]

File: src/almanack/metrics/data.py [hunk 2]
```diff
@@ -629,53 +645,128 @@ def _get_almanack_version() -> str:
...
+
+        except requests.HTTPError:
+            # Check for rate limit error (403 with a rate limit header)
+            if (
+                # ignore ruff linting code below as 403 is a known HTTP error code
+                response.status_code == 403  # noqa: PLR2004
+                and "X-RateLimit-Remaining" in response.headers
+            ):
+                if attempt < retries - 1:
+                    print(
```
[the remark this answers is not in the document]
Comment (d33bs) [reply src/almanack/metrics/data.py:697]: Thanks - I need to move these to the logger!

File: tests/metrics/test_data.py (hunk 1 above)
Comment (d33bs) [reply tests/metrics/test_data.py]: Thanks - I renamed this function to be a bit more broad based on your suggestion. Checking on the request failures / rate limit issues could be helpful but might get really complicated (these sometimes are taken care of by external dependencies). I'm going to plan on keeping things as simple as we can for now to help evolve the work as we go.

**Comment by d33bs** [comment]
Thanks @gwaybio for the review! After making a few changes based on your suggestions and bumping up the rate limit constraints for the GitHub requests I'm going to merge this in now.

**Pull request merged**